lib/dtutils/file.lua

```diff
@@ -0,0 +1,523 @@
+  Sections = {"Name", "Synopsis", "Description", "Return_Value"},
+  Name = [[check_if_file_exists - check if a file or path exist]],
+  Synopsis = [[local df = require "lib/dtutils.file"
+
+    local result = df.check_if_file_exists(filepath)
+      filepath - string - a file or path to check]],
+  Description = [[check_if_file_exists checks to see if a file or path exists]],
+  Return_Value = [[result - boolean - true if the file or path exists, false if it doesn't]],
+}
+
+function dtutils_file.check_if_file_exists(filepath)
+  local result = os.execute("test -e " .. filepath)
+  if not result then
+    result = false
+  end
+  return result
+end
+
+--[[
+  NAME
+    chop_filetype - remove a filetype from a filename
+
+  SYNOPSIS
+    local df = require "lib/dtutils.file"
+
+    local result = df.chop_filetype(path)
+      path - string - a filename with or without a path
+
+  DESCRIPTION
+    chop_filetype removes the filetype from the filename
+
+  RETURN VALUE
+    result - string - the path and filename without the filetype
+
+]]
+
+dtutils_file.libdoc.functions[#dtutils_file.libdoc.functions + 1] = {
+  Sections = {"Name", "Synopsis", "Description", "Return_Value"},
+  Name = [[chop_filetype - remove a filetype from a filename]],
+  Synopsis = [[local df = require "lib/dtutils.file"
+
+    local result = df.chop_filetype(path)
+      path - string - a filename with or without a path]],
+  Description = [[chop_filetype removes the filetype from the filename]],
+  Return_Value = [[result - string - the path and filename without the filetype]],
+}
+
+function dtutils_file.chop_filetype(path)
+  local length = dtutils_file.get_filetype(path):len() + 2
+  return string.sub(path, 1, -length)
+end
+
+--[[
+  NAME
+    file_copy - copy a file to another name/location
+
+  SYNOPSIS
+    local df = require "lib/dtutils.file"
+
+    local result = df.file_copy(fromFile, toFile)
+      fromFile - string - name of file to copy from
+      toFile - string - name of file to copy to
+
+  DESCRIPTION
+    copy a file using a succession of methods from operating system
+    to a pure lua solution
+
+  RETURN VALUE
+    result - boolean - nil on error, true on success
+
+]]
+
+dtutils_file.libdoc.functions[#dtutils_file.libdoc.functions + 1] = {
+  Sections = {"Name", "Synopsis", "Description", "Return_Value"},
+  Name = [[file_copy - copy a file to another name/location]],
+  Synopsis = [[local df = require "lib/dtutils.file"
+
+    local result = df.file_copy(fromFile, toFile)
+      fromFile - string - name of file to copy from
+      toFile - string - name of file to copy to]],
+  Description = [[copy a file using a succession of methods from operating system
+    to a pure lua solution]],
+  Return_Value = [[result - boolean - nil on error, true on success]],
+}
+
+function dtutils_file.file_copy(fromFile, toFile)
+  local result = nil
+  -- if cp exists, use it
+  if dtutils_file.check_if_bin_exists("cp") then
```

Comment:
This stemmed from using os.rename to move a file from $HOME/.local/tmp to wherever the collection was located.  When we crossed file system boundaries os.rename didn't work since it actually needed to do a copy, not just a rename.  I replaced os.rename with file_move() which attempts a rename and if that doesn't work then os.execture("mv..."), then if that fails  file_copy which tries to copy the file using os.execute("cp...") then finally a pure lua solution. So, in other words I'm trying to move the file using the least expensive method I can, then supplying alternatives in order of increasing expense.